### PR TITLE
UNDERTOW-2019: Blocking exchanges can set HEAD response length

### PR DESCRIPTION
**Pull request by carterkozak**
Jira: https://issues.redhat.com/browse/UNDERTOW-2019

### CODE REVIEW

File: core/src/main/java/io/undertow/io/UndertowOutputStream.java [hunk 1]
```diff
@@ -328,10 +329,14 @@ public void close() throws IOException {
         try {
             state |= FLAG_CLOSED;
             if (anyAreClear(state, FLAG_WRITE_STARTED) && channel == null) {
-                if (buffer == null) {
-                    exchange.getResponseHeaders().put(Headers.CONTENT_LENGTH, "0");
-                } else {
-                    exchange.getResponseHeaders().put(Headers.CONTENT_LENGTH, "" + buffer.position());
+                // Head requests may not write to the response, setting content-length directly instead
+                if (!(Methods.HEAD.equals(exchange.getRequestMethod())
+                        && exchange.getResponseHeaders().contains(Headers.CONTENT_LENGTH))) {
```
Comment (carterkozak) [inline core/src/main/java/io/undertow/io/UndertowOutputStream.java:334]: Alternatively we could use this path any time `Content-Length` is already set regardless of the method, although that could impact existing code if values are set incorrectly. What do you think?

Comment (berler) [reply core/src/main/java/io/undertow/io/UndertowOutputStream.java:335]: It looks like now this condition duplicates the isHeadRequestWithContentLength method.

Comment (carterkozak) [reply core/src/main/java/io/undertow/io/UndertowOutputStream.java]: 🤦 thanks! Looks like I pushed too soon!

File: core/src/test/java/io/undertow/server/handlers/HeadBlockingExchangeTestCase.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2014 Red Hat, Inc., and individual contributors
```
Comment (fl4via) [inline core/src/test/java/io/undertow/server/handlers/HeadBlockingExchangeTestCase.java:3]: thanks for your PR, @carterkozak !
Can you please change the Copyright year to 2022?

**Review by fl4via — CHANGES_REQUESTED** [review]

File: core/src/test/java/io/undertow/server/handlers/HeadBlockingExchangeTestCase.java (hunk 2 above)
Comment (carterkozak) [reply core/src/test/java/io/undertow/server/handlers/HeadBlockingExchangeTestCase.java]: Updated! Thanks for the review

**Review by fl4via — APPROVED** [review]

**Pull request merged**